File: app/src/main/java/io/github/summerdez/asmrplayer/data/remote/DlsiteRemoteFiles.java
package io.github.summerdez.asmrplayer.data.remote;

import android.text.TextUtils;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.InterruptedIOException;
import java.io.OutputStream;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Locale;
import java.util.Set;

final class DlsiteRemoteFiles {
    private DlsiteRemoteFiles() {
    }

    static String encodePath(String path) {
        if (TextUtils.isEmpty(path)) {
            return "";
        }
        String[] segments = path.split("/");
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < segments.length; i++) {
            if (i > 0) {
                builder.append('/');
            }
            builder.append(encodeQueryValue(segments[i]));
        }
        return builder.toString();
    }

    static String encodeQueryValue(String value) {
        try {
            return URLEncoder.encode(value == null ? "" : value, StandardCharsets.UTF_8.name())
                    .replace("+", "%20");
        } catch (IOException exception) {
            return "";
        }
    }

    static File localFileFor(File workDir, String relativePath) {
        String[] segments = relativePath == null ? new String[0] : relativePath.split("/");
        File current = workDir;
        for (String segment : segments) {
            String safe = safeFileName(segment);
            if (!safe.isEmpty()) {
                current = new File(current, safe);
            }
        }
        return current;
    }

    static File uniqueTarget(File targetFile, Set<String> usedTargets) {
        File candidate = targetFile;
        int index = 2;
        while (!usedTargets.add(candidate.getAbsolutePath())) {
            String name = targetFile.getName();
            int dot = name.lastIndexOf('.');
            String base = dot > 0 ? name.substring(0, dot) : name;
            String extension = dot > 0 ? name.substring(dot) : "";
            candidate = new File(targetFile.getParentFile(), base + " " + index + extension);
            index++;
        }
        return candidate;
    }

    static String safeFileName(String value) {
        String safe = TextUtils.isEmpty(value) ? "download" : value.trim();
        safe = safe.replaceAll("[\\\\/:*?\"<>|]+", "_");
        if (safe.equals(".") || safe.equals("..")) {
            return "download";
        }
        return safe;
    }

    static boolean looksLikeHtml(File file) throws IOException {
        if (file == null || !file.isFile()) {
            return false;
        }
        byte[] header = new byte[(int) Math.min(128, file.length())];
        try (FileInputStream input = new FileInputStream(file)) {
            int count = input.read(header);
            if (count <= 0) {
                return false;
            }
            String text = new String(header, 0, count, StandardCharsets.UTF_8).trim().toLowerCase(Locale.ROOT);
            return text.startsWith("<!doctype html") || text.startsWith("<html");
        }
    }

    static boolean looksLikeJson(File file) throws IOException {
        if (file == null || !file.isFile()) {
            return false;
        }
        byte[] header = new byte[(int) Math.min(128, file.length())];
        try (FileInputStream input = new FileInputStream(file)) {
            int count = input.read(header);
            if (count <= 0) {
                return false;
            }
            String text = new String(header, 0, count, StandardCharsets.UTF_8).trim();
            return text.startsWith("{") || text.startsWith("[");
        }
    }

    static String readTextFile(File file) throws IOException {
        StringBuilder builder = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(
                new FileInputStream(file), StandardCharsets.UTF_8))) {
            char[] buffer = new char[16 * 1024];
            int count;
            while ((count = reader.read(buffer)) != -1) {
                builder.append(buffer, 0, count);
            }
        }
        return builder.toString();
    }

    static void writeTextFile(File file, String text) throws IOException {
        try (OutputStream output = new FileOutputStream(file)) {
            output.write((text == null ? "" : text).getBytes(StandardCharsets.UTF_8));
        }
    }

    static void deleteQuietly(File file) {
        if (file != null && file.exists() && !file.delete()) {
            file.deleteOnExit();
        }
    }

    static String summarizeBody(String body) {
        if (body == null) {
            return "";
        }
        String text = body.replaceAll("\\s+", " ").trim();
        return text.length() <= 160 ? text : text.substring(0, 160);
    }

    static void throwIfInterrupted() throws InterruptedIOException {
        if (Thread.currentThread().isInterrupted()) {
            throw new InterruptedIOException("下载已取消");
        }
    }
}

File: app/src/main/java/io/github/summerdez/asmrplayer/data/remote/DlsiteRemoteFiles.kt
package io.github.summerdez.asmrplayer.data.remote

import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStreamReader
import java.io.InterruptedIOException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.Locale
import java.util.regex.Pattern

object DlsiteRemoteFiles {
    private val pathSeparator: Pattern = Pattern.compile("/")

    fun encodePath(path: String?): String {
        if (path.isNullOrEmpty()) {
            return ""
        }
        val segments = pathSeparator.split(path)
        val builder = StringBuilder()
        for (i in segments.indices) {
            if (i > 0) {
                builder.append('/')
            }
            builder.append(encodeQueryValue(segments[i]))
        }
        return builder.toString()
    }

    fun encodeQueryValue(value: String?): String {
        return try {
            URLEncoder.encode(value ?: "", StandardCharsets.UTF_8.name())
                .replace("+", "%20")
        } catch (exception: IOException) {
            ""
        }
    }

    fun localFileFor(workDir: File?, relativePath: String?): File? {
        val segments = if (relativePath == null) emptyArray() else pathSeparator.split(relativePath)
        var current = workDir
        for (segment in segments) {
            val safe = safeFileName(segment)
            if (safe.isNotEmpty()) {
                current = File(current, safe)
            }
        }
        return current
    }

    fun uniqueTarget(targetFile: File?, usedTargets: MutableSet<String>?): File {
        val source = targetFile ?: throw NullPointerException("targetFile")
        val targets = usedTargets ?: throw NullPointerException("usedTargets")
        var candidate = source
        var index = 2
        while (!targets.add(candidate.absolutePath)) {
            val name = source.name
            val dot = name.lastIndexOf('.')
            val base = if (dot > 0) name.substring(0, dot) else name
            val extension = if (dot > 0) name.substring(dot) else ""
            candidate = File(source.parentFile, "$base $index$extension")
            index++
        }
        return candidate
    }

    fun safeFileName(value: String?): String {
        var safe = if (value.isNullOrEmpty()) "download" else value.trim { it <= ' ' }
        safe = safe.replace(Regex("[\\\\/:*?\"<>|]+"), "_")
        if (safe == "." || safe == "..") {
            return "download"
        }
        return safe
    }

    @Throws(IOException::class)
    fun looksLikeHtml(file: File?): Boolean {
        if (file == null || !file.isFile) {
            return false
        }
        val header = ByteArray(minOf(128L, file.length()).toInt())
        FileInputStream(file).use { input ->
            val count = input.read(header)
            if (count <= 0) {
                return false
            }
            val text = String(header, 0, count, StandardCharsets.UTF_8)
                .trim { it <= ' ' }
                .lowercase(Locale.ROOT)
            return text.startsWith("<!doctype html") || text.startsWith("<html")
        }
    }

    @Throws(IOException::class)
    fun looksLikeJson(file: File?): Boolean {
        if (file == null || !file.isFile) {
            return false
        }
        val header = ByteArray(minOf(128L, file.length()).toInt())
        FileInputStream(file).use { input ->
            val count = input.read(header)
            if (count <= 0) {
                return false
            }
            val text = String(header, 0, count, StandardCharsets.UTF_8).trim { it <= ' ' }
            return text.startsWith("{") || text.startsWith("[")
        }
    }

    @Throws(IOException::class)
    fun readTextFile(file: File?): String {
        val builder = StringBuilder()
        BufferedReader(InputStreamReader(FileInputStream(file), StandardCharsets.UTF_8)).use { reader ->
            val buffer = CharArray(16 * 1024)
            while (true) {
                val count = reader.read(buffer)
                if (count == -1) {
                    break
                }
                builder.append(buffer, 0, count)
            }
        }
        return builder.toString()
    }

    @Throws(IOException::class)
    fun writeTextFile(file: File?, text: String?) {
        FileOutputStream(file).use { output ->
            output.write((text ?: "").toByteArray(StandardCharsets.UTF_8))
        }
    }

    fun deleteQuietly(file: File?) {
        if (file != null && file.exists() && !file.delete()) {
            file.deleteOnExit()
        }
    }

    fun summarizeBody(body: String?): String {
        if (body == null) {
            return ""
        }
        val text = body.replace(Regex("\\s+"), " ").trim { it <= ' ' }
        return if (text.length <= 160) text else text.substring(0, 160)
    }

    @Throws(InterruptedIOException::class)
    fun throwIfInterrupted() {
        if (Thread.currentThread().isInterrupted) {
            throw InterruptedIOException("下载已取消")
        }
    }
}
